### PR TITLE
Implementing Debug-friendly mode for local optimizer and enabling it in Debug

### PR DESCRIPTION
**Pull request by VSadov**
Implementing Debug-friendly mode for local optimizer and enabling it in Debug

This is a step towards #4103

### CODE REVIEW

**Comment by VSadov** [comment]
@AlekseyTs @jaredpar @gafter @TyOverby @agocke @tmat - please review

**Comment by VSadov** [comment]
Once we have debug-freindly optimization mode, it makes sense to use it in Debug and thus it is useful (and testable) by itself before completing the whole fix for 4103. Therefore I want this to be a separate change.

**Comment by jaredpar** [comment]
CC @jacdavis as well.

**Comment by VSadov** [comment]
The end result of this change will be that short-lived temps (like introduced by x++) will be optimized even in Debug. However the optimizer is not supposed to perform cross-statement optimizations and is not supposed to elide user-defined locals (or any long-lived locals) since that would be a problem for EnC.

The other part of the fix for #4103 would be plumbing through appropriate switches and, based on that, using this same mode conditionally in Release builds.

**Comment by jacdavis** [comment]
Looks good. Vlad, will plumbing this through the project system be part of the next change?

File: src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs [hunk 1]
```diff
@@ -294,9 +296,10 @@ internal enum ExprContext
     //
     internal class StackOptimizerPass1 : BoundTreeRewriter
     {
-        private int _counter;
+        private readonly bool _debugFriendly;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs]: Consider adding a comment about what does debugFriendly mean exactly?

**Comment by AlekseyTs** [comment]
Do we need to do similar change for VB?

**Comment by AlekseyTs** [comment]
LGTM

**Comment by VSadov** [comment]
@AlekseyTs - yes 4103 applies to VB as well, so similar changes will be done to VB too.

File: src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs (hunk 1 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs]: Will add in next change.

**Comment by VSadov** [comment]
@jacdavis - yes, the next part would be to enable this conditionally in Release and plumb through whatever switch we agree on.

**Comment by tmat** [comment]
:+1:

File: src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs [hunk 2]
```diff
@@ -1422,6 +1435,15 @@ private void RecordLabel(LabelSymbol label)
             }
         }
 
+        private void ShouldNotSchedule(LocalSymbol localSymbol)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs]: `ShouldNotScheduleToStack`?

File: src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs [hunk 3]
```diff
@@ -1525,9 +1547,10 @@ private void RecordVarWrite(LocalSymbol local)
             locInfo.LocalDefs.Add(locDef);
         }
 
-        private static bool CanScheduleToStack(LocalSymbol local)
+        private bool CanScheduleToStack(LocalSymbol local)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs]: I'm confused about this terminology. What does it mean to "schedule to stack"? Is there any literature on this optimization?

**Comment by agocke** [comment]
:+1:

**Pull request merged**